src/Plugin/QuickForm/SaatenUnionSpraying.php

```diff
@@ -0,0 +1,746 @@
+      'title' => $this->t('Speed Driven'),
+      'description' => $this->t('The travelling speed when spraying, where relevant.'),
+      'measure' => ['#value' => 'ratio'],
+      'units' => ['#options' => $speed_driven_units_options],
+    ]);
+
+
+    $plant_asset_options = $this->getPlantAssetOptions();
+    $form['plant_asset'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Planting asset'),
+      '#description' => $this->t('The plant asset that this log relates to.'),  
+      '#options' => $plant_asset_options,
+      '#required' => TRUE,
+      '#multiple' => TRUE,
+    ];
+
+    $equipment_options = $this->getEquipmentOptions();
+    $form['equipment'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Equipment'),
+      '#description' => $this->t('Select the tractor, sprayer and nozzle equipment assets used for this spraying'),  
+      '#options' => $equipment_options,
+      '#required' => TRUE,
+      '#multiple' => TRUE,
+    ];
+
+    $form['justification_target'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Justification/Target'),
+      '#description' => $this->t('The reason the operation is necessary, and any target pest(s) where applicable.'),
+      '#required' => TRUE,
+    ];
+
+    $form['plant_growth_stage'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Plant growth stage'),
+      '#description' => $this->t('The plant growth stage when the product was applied.'),
+      '#required' => FALSE,
+    ];
+
+    $form['equipment_rinsed'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Equipment triple rinsed'),
+      '#description' => $this->t('Select if the equipment was triple rinsed after the job was completed.'),
+      '#return_value' => 'Yes',
+    ];
+
+    $form['equipment_clear'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Equipment All Clear'),
+      '#description' => $this->t('Select if the equipment all clear after the job was completed.'),
+      '#return_value' => 'Yes',
+    ];
+
+    $form['equipment_washed'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Equipment Clear Washed'),
+      '#description' => $this->t('Select if the equipment was clear washed after the job was completed.'),
+      '#return_value' => 'Yes',
+    ];
+
+    // Log notes.
+    $form['notes'] = [
+      '#type' => 'textarea',
+      '#title' => $this->t('Notes'),
+      '#description' => $this->t('Any additional notes.'),
+      '#weight' => 20,
+    ];
+    
+    return $form;
+  }
+
+  /**
+  * Submit handler for Saaten Union Spraying Quick Form.
+  */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+
+  // Quantities.
+  $quantity_keys = [
+    'product_rate', 
+    'area',
+    'temperature',
+    'wind_speed',
+    'water_volume',
+    'total_product_quantity',
+    'pressure',
+    'speed_driven'
+  ];
+  $quantities = $this->getQuantities($quantity_keys, $form_state);
```

Comment:
We will need to do additional processing on the `product_rate` and `total_product_quantity` quantities:

- Both of these need to set the `quantity.material_type` field to the material type of the selected `material` asset.
- Additionally the `total_product_quantity` needs to be an inventory decrement referencing the selected `material` asset.

I think you'll want something like this:
```php
$material = $asset_storage->load($form_state->getValue('product'));
$quantities['product_rate']['material_type'] = $material->get('material_type')->value;
$quantities['total_product_quantity']['material_type'] = $material->get('material_type')->value;
$quantities['total_product_quantity']['inventory_adjustment'] = 'decrement';
$quantities['total_product_quantity']['inventory_asset'] = $material;

Reply:
Actually you can't reference `$quantities['product_rate']` because it is a numerical array. Thus, I would build these quantities first, then get all the other simple quantities.

```php
$product_rate = $form_state->getValue('product_rate');
$product_rate['material_type'] = $material->get('material_type')->value;

$quantities[] = $product_rate;
$quantities[] = $total_product_quantity;
$quantities += $this->getQuantities($quantity_keys, $form_state);